File: multi-gammon-core/src/main/java/org/akoshterek/backgammon/bearoff/BearoffContext.java
package org.akoshterek.backgammon.bearoff;

import com.google.common.io.LittleEndianDataInputStream;
import org.apache.commons.io.IOUtils;

import java.io.IOException;

/**
 * @author Alex
 *         date 31.08.2015.
 */
public class BearoffContext {
    private boolean isTwoSided; /* type of bearoff database */
    private int nPoints;    /* number of points covered by database */
    private int nChequers;  /* number of chequers for one-sided database */
    private String szFilename; /* filename */
    /* one sided dbs */
    private boolean fCompressed; /* is database compressed? */
    private boolean fGammon;     /* gammon probs included */

    private boolean fND;         /* normal distibution instead of exact dist? */
    /* two sided dbs */
    private boolean fCubeful;    /* cubeful equities included */
    private byte[] data;        /* pointer to data in memory */

    public static BearoffContext bearoffInit (String szFilename) {
        BearoffContext pbc = new BearoffContext();
        byte[] sz = new byte[40];

        pbc.szFilename = szFilename;

        try (LittleEndianDataInputStream inputStream = new LittleEndianDataInputStream(BearoffContext.class.getResourceAsStream(szFilename))) {
	        // read header
            inputStream.readFully(sz);
            // detect bearoff program
            String header = new String(sz, 0, 5, "UTF-8");
            if(!"gnubg".equals(header)) {
                throw new IllegalArgumentException("Unknown bearoff database");
            }

        	// one sided or two sided?
            String type = new String(sz, 6, 2, "UTF-8");
            if("TS".equals(type)) {
                pbc.isTwoSided = true;
            } else if("OS".equals(type)) {
                pbc.isTwoSided = false;
            } else {
                throw new IllegalArgumentException("Illegal bearoff type " + type);
            }

		    // number of points
            String pointsStr = new String(sz, 9, 2, "UTF-8");
            pbc.nPoints = Integer.valueOf(pointsStr);
            if (pbc.nPoints < 1 || pbc.nPoints >= 24) {
                throw new IllegalArgumentException("illegal number of points " + pbc.nPoints);
            }

		    // number of chequers
            String nChequersStr = new String(sz, 12, 2, "UTF-8");
            pbc.nChequers = Integer.valueOf(nChequersStr);
            if (pbc.nChequers < 1 || pbc.nChequers > 15) {
                throw new IllegalArgumentException("illegal number of chequers " + pbc.nChequers);
            }

            if (pbc.isTwoSided) {
                // options for two-sided dbs
                pbc.fCubeful = Integer.valueOf(new String(sz, 15, 1, "UTF-8")) != 0;
            } else {
		        // options for one-sided dbs
                pbc.fGammon = Integer.valueOf(new String(sz, 15, 1, "UTF-8")) != 0;
                pbc.fCompressed = Integer.valueOf(new String(sz, 17, 1, "UTF-8")) != 0;
                pbc.fND = Integer.valueOf(new String(sz, 19, 1, "UTF-8")) != 0;
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        readBinaryData(pbc, szFilename);
        return pbc;
    }

    public void readBearoffData(int offset, byte[] buf, int nBytes) {
        System.arraycopy(data, offset, buf, 0, nBytes);
    }

    private static void readBinaryData(BearoffContext pbc, String szFilename) {
        try {
            pbc.data = IOUtils.toByteArray(BearoffContext.class.getResourceAsStream(szFilename));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public int getnPoints() {
        return nPoints;
    }

    public int getnChequers() {
        return nChequers;
    }

    public boolean isfCubeful() {
        return fCubeful;
    }

    public boolean isTwoSided() {
        return isTwoSided;
    }

    public boolean isfND() {
        return fND;
    }

    public boolean isfGammon() {
        return fGammon;
    }

    public boolean isfCompressed() {
        return fCompressed;
    }

    public String getSzFilename() {
        return szFilename;
    }
}

File: multi-gammon-core/src/main/java/org/akoshterek/backgammon/bearoff/BearoffContext.scala
package org.akoshterek.backgammon.bearoff

import java.io.{IOException, InputStream}

import com.google.common.io.LittleEndianDataInputStream
import org.apache.commons.io.IOUtils
import resource.managed

/**
  * @author Alex
  *         date 31.08.2015.
  * @param isTwoSided type of bearoff database
  * @param points number of points covered by database
  * @param chequers number of chequers for one-sided database
  * @param isWithGammonProbs gammon probabilities included
  * @param isCubeful cubeful equities included
  * @param isCompressed is database compressed?
  * @param isNormalDistribution normal distribution instead of exact dist?
  * @param data pointer to data in memory
  */
final class BearoffContext private(val isTwoSided: Boolean,
                             val points: Int,
                             val chequers: Int,
                             val isWithGammonProbs: Boolean,
                             val isCubeful: Boolean,
                             val isCompressed: Boolean,
                             val isNormalDistribution: Boolean,
                             private val data: Array[Byte]) {

    def readBearoffData(offset: Int, nBytes: Int): Array[Byte] = {
      data.slice(offset, offset + nBytes)
    }
}

object BearoffContext {
  def apply(szFilename: String): BearoffContext = {
    managed(new LittleEndianDataInputStream(classOf[BearoffContext].getResourceAsStream(szFilename))).acquireAndGet(inputStream => {
      loadBearoffDatabase(inputStream)
    })
  }

  @throws[IOException]
  private def loadBearoffDatabase(inputStream: InputStream): BearoffContext = {
    // read header
    val data = IOUtils.toByteArray(inputStream)
    // detect bearoff program
    require("gnubg".equals(new String(data, 0, 5, "UTF-8")), "Unknown bearoff database")

    // one sided or two sided?
    val _type = new String(data, 6, 2, "UTF-8")
    val isTwoSided = _type match {
      case "TS" => true
      case "OS" => false
      case _ => throw new IllegalArgumentException("Illegal bearoff type " + _type)
    }

    // number of points
    val pointsStr = new String(data, 9, 2, "UTF-8")
    val nPoints = Integer.valueOf(pointsStr)
    require (nPoints >= 1 || nPoints < 24, "illegal number of points " + nPoints)

    // number of chequers
    val nChequersStr = new String(data, 12, 2, "UTF-8")
    val nChequers = Integer.valueOf(nChequersStr)
    require(nChequers >= 1 && nChequers <= 15, "illegal number of chequers " + nChequers)

    val fCubeful = if (isTwoSided) Integer.valueOf(new String(data, 15, 1, "UTF-8")) != 0 else false
    val fGammon = if (isTwoSided) false else Integer.valueOf(new String(data, 15, 1, "UTF-8")) != 0
    val fCompressed = if (isTwoSided) false else Integer.valueOf(new String(data, 17, 1, "UTF-8")) != 0
    val fND = if (isTwoSided) false else Integer.valueOf(new String(data, 19, 1, "UTF-8")) != 0

    new BearoffContext(isTwoSided, nPoints, nChequers, fGammon, fCubeful, fCompressed, fND, data)
  }
}
